Engineering/Frontend/A Fragment Colocation Pattern with React & Apollo GraphQL.md

```diff
@@ -0,0 +1,266 @@
+    weight
+  }
+  ${ANIMAL_FRAGMENT}
+`;
+
+```
+
+By this point, we are still missing something until we can build the `CatCard` and `DogCard` components - the Typescript types.
+
+With `@graphql-codegen/cli`, we can convert these fragments into Typescript interfaces by running a CLI script. I will not go into details into how the tool work so you should also give [this link]([Home – GraphQL Code Generator (the-guild.dev)](https://the-guild.dev/graphql/codegen)) a look - they provide an interactive example. 
+
+Basically `@graphql-codegen/cli` will:
+- Call our GraphQL backend to fetch the schema
+- Scan our codebase for query, mutation and fragment definitions to convert
+- Put the output into a file that we have specificed
+
+After running the CLI, the typing output will look like below:
+
+```ts
+// Output file: graphql/generated.ts
+
+export type AnimalFragment {
+	__typename?: 'AnimalModel';
+	id: Scalars['String'];
+	name: Scalars['String'];
+	bread: Scalars['String'];
+}
+
+export type CatFragment {
+	__typename?: 'AnimalModel';
+	id: Scalars['String'];
+	name: Scalars['String'];
+	bread: Scalars['String'];
+	age: Scalars['Int'];
+}
+
+export type AnimalFragment {
+	__typename?: 'AnimalModel';
+	id: Scalars['String'];
+	name: Scalars['String'];
+	bread: Scalars['String'];
+	weigth: Scalars['Int'];
+}
+```
+
+Now that we have everything we need, let's build the `CatCard` and `DogCard` components:
+
+```ts
+import { CatFragment } from 'graphql/generated';
+
+// const ANIMAL_FRAGMENT = gql`...`
+
+// const CAT_FRAGMENT = gql`...`
+
+// const DOG_FRAGMENT = gql`...`
+
+const CatCard = (props: { cat: CatFragment }) => {
+	const { cat } = props;
+
+	// Component rendering logic
+}
+
+const DogCard = (props: { cat: DogFragment }) => {
+	const { cat } = props;
+
+	// Component rendering logic
+}
+```
+
+The properties `cat` and `dog` will have the types we have defined for the fragments they are actually using - an exact map from GraphQL models to Typescript types that we can be sure will always be accurate as long as the fragments we define match the schema from GraphQL backend.
+
+Next, let's build the `CatList` and `DogList` component and see how we handle queries. Let's defined 2 queries to get cats and dogs:
+
+```ts
+// ... import needed stuff
+import { gql } from '@apollo/client';
+ 
+gql`
+    query GetCatList {
+      cats {
+        ...CatFragment
+      }
+    }
+    ${CAT_FRAGMENT}
+`
+ 
+gql`
+    query GetDogList {
+      dogs {
+        ...DogFragment
+      }
+    }
+    ${DOG_FRAGMENT}
+`
+```
+
+Then we run  `@graphql-codegen/cli` again. Depending on how we set-up the CLI, output will vary so the below are what I normally work with:
+- `useGetCatListQuery`: A hook that fires a request to fetch cat list & return the data.
+- `useGetCatListQueryLazy`: A hook that returns a function to fetch cat list in case we want to manually get the list
+- A variety of Typescript types for query document, variables or return result
```

Comment:
```suggestion
- `useGetCatListQueryLazy`: A hook that returns a function to fetch cat list in case we want to manually get the list.
- A variety of Typescript types for query document, variables or return result.
```